Grant/contracts/BuilderGrant.sol

```diff
@@ -0,0 +1,350 @@
+        uint16 totalClaimed; // total claimed mini building count for the season
+        Grant[3] grants;
+        Ranker[] rankers;
+    }
+    Season[] private seasons;
+
+    event Deposit(address indexed depositor, uint256 huntAmount);
+    event SetSeasonData(uint256 indexed seasonId, uint256 rankersCount, uint16[3] grantsAmount);
+    event EmergencyWithdraw(address indexed withdrawer, uint256 huntAmount);
+    event ClaimByTop3(
+        address indexed claimer,
+        uint256 indexed seasonId,
+        uint256 ranking,
+        uint8 claimType,
+        uint16 countForSelf,
+        uint16 countForDonation
+    );
+    event ClaimDonation(address indexed claimer, uint256 indexed seasonId, uint256 ranking, uint256 amount);
+
+    constructor(address bond, address hunt, address miniBuilding) Ownable(msg.sender) {
+        BOND = IMCV2_Bond(bond); // base: 0xc5a076cad94176c2996B32d8466Be1cE757FAa27
+        HUNT = IERC20(hunt); // base: 0x37f0c2915CeCC7e977183B8543Fc0864d03E064C
+        MINI_BUILDING_ADDRESS = miniBuilding; // base: 0x475f8E3eE5457f7B4AAca7E989D35418657AdF2a
+
+        // gas saving - approve infinite HUNT to Bond for minting
+        HUNT.approve(bond, type(uint256).max);
+    }
+
+    modifier validSeasonId(uint256 seasonId) {
+        if (seasonId >= seasons.length) revert InvalidSeasonId();
+        _;
+    }
+
+    function currentSeason() external view returns (uint256) {
+        return seasons.length; // The last claimable season = length - 1
+    }
+
+    function deposit(uint256 huntAmount) external {
+        address msgSender = _msgSender();
+        if (!HUNT.transferFrom(msgSender, address(this), huntAmount)) revert TokenTransferFailed();
+
+        emit Deposit(msgSender, huntAmount);
+    }
+
+    function emergencyWithdraw() external onlyOwner {
+        address msgSender = _msgSender();
+        uint256 balance = HUNT.balanceOf(address(this));
+        if (balance == 0) revert NothingToWithdraw();
+        if (!HUNT.transfer(msgSender, balance)) revert TokenTransferFailed();
+
+        emit EmergencyWithdraw(msgSender, balance);
+    }
+
+    function setSeasonData(
+        uint256 seasonId,
+        uint16[3] calldata grantsAmount,
+        uint48[] calldata fids,
+        address[] calldata wallets
+    ) external onlyOwner {
+        // Validate params
+        if (fids.length == 0 || fids.length != wallets.length) revert InvalidRankersParams();
+
+        // Check if we have enough HUNT to mint all Mini Building NFTs
+        if (
+            HUNT.balanceOf(address(this)) <
+            HUNT_PER_MINI_BUILDING * (grantsAmount[0] + grantsAmount[1] + grantsAmount[2])
+        ) revert NotEnoughGrantBalance();
+
+        // Check if there are enough ranker data is provided if the 1st ranker donates 100%
+        // e.g. If the top grant is 10,000 HUNT, include up to 103 rankers to allow donations to ranks 4-103
+        if (grantsAmount[0] > (fids.length - 3)) revert InvalidGrantAmount();
```

Comment:
1. LOW: There's an implicit assumption here that `grantsAmount[0] >= grantsAmount[1] >= grantsAmount[2]`, maybe add this as a validation?

2. INFO: This constrains the `grantAmount` to at most the number of tip recipients in the season; is that always desirable?

Reply:
Nice catch!
Yes, we always want the largest amount in `grantAmount[0]` because it represents the grant for the first ranker.

I will add some validations.